README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Eva-Scan

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![Codecov test
coverage](https://codecov.io/gh/JMPivette/evavelo/branch/master/graph/badge.svg)](https://codecov.io/gh/JMPivette/evavelo?branch=master)

<!-- badges: end -->

## Résumé

Le package `evavelo` applique des traitements automatiques sur les
fichiers d’enquêtes liés à la méthode
[EVA-VELO](https://www.velo-territoires.org/ressources/categorie/publication-reference/?resource-id=18202#resource-eva-velo)
(Méthode nationale pour l’évaluation des retombées des véloroutes)
développée par [**Vélo et
Territoires**](https://www.velo-territoires.org/)

Ce projet est composé de fonctions de traitements de fichier mais aussi
d’une interface web.

## Installation

``` r
# Development version from GitHub
# install.packages("devtools")
devtools::install_github("JMPivette/evavelo")
```

## Traitement d’enquête Eva-Velo

La function principale `process_evavelo()` permet de traiter un fichier
d’enquête xlsx avec la méthode Eva-Vélo. Cette function va notamment:

-   Vérifier l’intégrité du fichier
-   Remplir le champ `categorie_corrige`
-   Calculer les différentes distances par rapport au point d’enquête
-   Trouver les codes COG des communes de résidence et d’hébergement
-   Remplir les champs `id_section_origine` et `id_section_dest`

``` r
xlsx_path <- system.file("example-data/02_simplified.xlsx", package = "evavelo")

## Process File
eva_data_processed <- evavelo::process_evavelo(xlsx_path)

## Read outputs
lapply(eva_data_processed, head)
#> $comptages_man_post_traitements
#>   id_quest categorie_visuelle_cycliste_corrige
#> 1     <NA>                                <NA>
#> 2     <NA>                              Loisir
#> 3     <NA>                                <NA>
#> 4     <NA>                              Loisir
#> 5   106aA1                             Sportif
#> 6     <NA>                                <NA>
#> 
#> $enquetes_post_traitement
#>   id_quest categorie_corrige distance_domicile_enq distance_dom_enq_reelle
#> 1   106aA1           Sportif             0.5241257               0.7297089
#> 2   106aA2        Utilitaire                    NA                      NA
#> 3   106aA3           Sportif             5.0566412               6.7403990
#> 4   106aA4            Loisir             0.5241257               0.7297089
#> 5   106aA5           Sportif                    NA                      NA
#> 6   106aA6            Loisir           156.9983854             172.7165825
#>   distance_dom_enq_reelle_regions distance_dom_enq_reelle_france
#> 1                       0.7297089                              0
#> 2                              NA                             NA
#> 3                       6.7403990                              0
#> 4                       0.7297089                              0
#> 5                              NA                             NA
#> 6                     172.7165825                              0
#>   distance_dom_enq_reelle_etranger distance_heb_enq distance_heb_enq_reelle
#> 1                                0               NA                      NA
#> 2                               NA               NA                      NA
#> 3                                0               NA                      NA
#> 4                                0               NA                      NA
#> 5                               NA               NA                      NA
#> 6                                0               NA                      NA
#>   iti_dep_iti_valide id_section_origine iti_arr_iti_valide id_section_dest
#> 1               <NA>                 NA               <NA>              NA
#> 2               <NA>                 NA               <NA>              NA
#> 3               <NA>                 NA               <NA>              NA
#> 4               <NA>                 NA               <NA>              NA
#> 5               <NA>                 NA               <NA>              NA
#> 6               <NA>                 NA               <NA>              NA
#>   ville_heb_cog_lau ville_res_cog_lau
#> 1              <NA>             17300
#> 2              <NA>              <NA>
#> 3              <NA>             17028
#> 4              <NA>             17300
#> 5              <NA>              <NA>
#> 6              <NA>             33318
```

## Traitement étapes par étapes

Il est possible de décomposer tous les traitements effectués par
`process_evavelo()` étape par étape:

### Lecture du fichier

La fonction `read_evavelo()` permet de lire un fichier xlsx et de
récupérer un objet avec toutes les informations intéressantes pour le
traitement.

``` r
eva_data <- evavelo::read_evavelo(xlsx_path)
```

### Géocodage des villes

La fonction `geocode_evavelo()` permet de modifier l’objet en géocodant
les noms de communes

``` r
eva_data_geocoded <- evavelo::geocode_evavelo(eva_data)
#> Vérification des noms de communes
#> ---------------------------------
#> ...Vérification de table_communes.............
#> Les communes suivantes n'existent plus et doivent etre remplacées par les communes nouvelles:
#>  Hiers-Brouage(17189) -> Marennes-Hiers-Brouage(17219)
#>  Marennes(17219) -> Marennes-Hiers-Brouage(17219)
#>  Locmaria-Berrien(29129) -> Poullaouen(29227)
#>  Château-d'Olonne(85060) -> Les Sables-d'Olonne(85194)
#>  Olonne-sur-Mer(85166) -> Les Sables-d'Olonne(85194)
#> Impossible de reconnaître les communes suivantes:
#>  Lanouée(56102)
#> 
#> ...Vérification de ville_res.............
#> Les villes suivantes ont ete ignorées. Propositions de corrections:
#>  Angoulins-sur-Mer (17890) ->    Angoulins (17690)
#>  Angoulins-sur-Mer (17691) ->    Angoulins (17690)
#>  Angoulins-sur-Mer (17960) ->    Angoulins (17690)
#>  Angoulins-sur-Mer (17690) ->    Angoulins (17690)
#>  Arboras (34151) ->  Arboras (34150)
#>  Avallon (89201) ->  Avallon (89200)
#>  Avanton (86171) ->  Avanton (86170)
#>  Cachan (94231) ->   Cachan (94230)
#>  Chatelaillon (17340) ->     Châtelaillon-Plage (17340)
#>  Fouras (17451) ->   Fouras (17450)
#>  Le Mans (72700) ->  Le Mans (72100)
#>  Plouénan (39420) ->     Plouénan (29420)
#>  Rennes (35160) ->   Rennes (35000)
#>  Rivedoux (17940) ->     Rivedoux-Plage (17940)
#>  Rivedoux-Plage (17590) ->   Rivedoux-Plage (17940)
#>  Saint-Germain-du-Puy (18391) ->     Saint-Germain-du-Puy (18390)
#>  Saint-Joseph (97980) ->     Saint-Joseph (97480)
#>  Saint-Pol-de-Léon (22250) ->    Saint-Pol-de-Léon (29250)
#> Interprétation de noms de communes étrangères:
#>  Chine : Benjing -> Beijing
#> Villes inconnues:
#>  Noor mout (Pays-Bas)
#> 
#> ...Vérification de ville_heb.............
#> Impossible de trouver les communes suivantes:
#>   Camping-municipal-henvic
#>   Chaix
#>   Île de Ré
#> Interpretation de communes mal nommées:
#>  Chatelaillon -> Châtelaillon-Plage (17094) 
#>  Saint-Pol -> Saint-Pol-de-Léon (29259)
#> 
#> ...Vérification de iti_depart_itineraire.............
#> Impossible de trouver les communes suivantes:
#>   Uméa
#> 
#> ...Vérification de iti_arrivee_itineraire.............
#> Impossible de trouver les communes suivantes:
#>   Dortmund
#> 
#> ...Vérification de nom_site_enq.............
```

### Process

La fonction `process_evavelo()` que nous avons déjà utilisé sur un
fichier Excel peut aussi être utilisée directement sur un objet
précédemment géocodé avec `geocode_evavelo()`

``` r
eva_data_processed <- evavelo::process_evavelo(eva_data_geocoded)
#> Calcul des distances...
#> Les villes de départ d'itinéraire suivantes sont trop éloignées de l'itinéraire (>30km):
#>  St Brieuc (38A23)
#>  
#> Les villes d'arrivée d'itinéraire suivantes sont trop éloignées de l'itinéraire (>30km):
#>  Paris (38A34)
#>  
#> Liste des régions de l'itinéraire à partir de tables_communes:
#>  Bretagne, Nouvelle-Aquitaine, Pays de la Loire
#> 
#> Correction de catégories pour 17 questionnaires ....
#> Il n'a pas été possible de corriger les catégories de 3 questionnaire(s).
#>  La catégorie du déclarant sera utilisée:
#>  141A35, 38A45, 38A46
```

## Interface Web

La fonction `evavelo::app_run()` permet d’utiliser le package via une
interface web:

![UI Screenshot](man/figures/eva-scan-ui.png?raw=true "Title")

## Classification de compteurs similaires

Le package `eva-scan` permet aussi de créer une classification
hiérarchique des compteurs à partir de données de comptages
automatiques. Cette classification est disponible dans l’interface Web.


NEWS.md
# evavelo 1.2.1

- Increase input file maximum size to 500MB

# evavelo 1.2.0

- Clustering: don't allow Dendogram visualisation above 50 sites. (otherwise rendering is way too slow)

# evavelo 1.1.5

- Fix bug in {openxlsx} when convering date-time on long object

# evavelo 1.1.4

- Fix a bug in distance calculation in 1.1.3 that was released too quickly. (bug introduced with latest version of {sf})

# evavelo 1.1.3

- Change max upload size from 30MB to 50MB

# evavelo 1.1.2

- Update README

# evavelo 1.1.1

- Allow `categorie_visuelle_cycliste` with plural version (Utilitaires instead of Utilitaire for example) #79

- Fix #81 to improve information on logs displayed in case of duplicated ids

- Fix #84 for better mismatch detection based on 'id_quest' value.

# evavelo 1.1.0

- Store correction proposition during geocoding phase. These proposition were only displayed in the log before. They now appear in new columns called "proposition_*" (Fix #65 #76)

- Fix #77 Missing data in one of the input worksheets now creates an error. Empty worksheet `comptages_automatiques` is allowed but creates a warning.

- Fix #75 that displayed a progress bar in the logs

- Fix a bug on check_evavelo() when a column was missing from *_post_traitement version of one of the sheets (linked to #78)

# evavelo 1.0.3

- Remove dependency to dev version of openxlsx (following release of 4.2.4)

- Fix bug when trying to display cluster graph without clicking on the button

- Fix bug when importing a file with empty "enquetes" sheet.

# evavelo 1.0.2

- Update documentation

- add examples as exported objects

# evavelo 1.0.1

- Update documentation

# evavelo 1.0.0

- Read counting data from worksheet `comptages_automatiques` 

- Add site clustering and dendogram visualization #24

# evavelo 0.8.1

- Improve UI #66

# evavelo 0.8.0

- Segmentation of `distance_dom_enq_reelle` in 3 fields `distance_dom_enq_reelle_regions` `distance_dom_enq_reelle_france` `distance_dom_enq_reelle_etranger` #63

- Detects foreign cities in `ville_res` #57. We use Open Street Map engine (Nomatim) that has a rate limit of 1 request per second. Geocoding can be slow if there are a lot of different cities outside of France. 

- Speed improvement in geocoding by using local list of cities (France and Outside)

# evavelo 0.7.1

- Detect outliers on tour_dep_transp in `check_num_outliers()` #59

# evavelo 0.7.0

- Improve documentation 

- Add a pkgdown website for documentation

# evavelo 0.6.0

- Add find_wrong_geocoding() to export all the cities that couldn't be geocoded.

- Improve UI to separate each treatment

- Add check_num_oultiers() function (#34)

- Exploratory analysis on outliers #34

# evavelo 0.5.1

- Add volume and `taille_groupe` check on multiple id_quest #53

- Detects potential multiple answers #54

# evavelo 0.5.0

- User Interface improvement #33

- Logs are now all written in french #20

# evavelo 0.4.2

- Fix issues #50 #51 #49 on categories.

- #39 : change `coherence` criteria

# evavelo 0.4.1

## Bug Fixes 

- Fix issue #37 that didn't compare correctly id_quest from `enquete` and `comptage`

- Fix issue #40 where categories were not update in `comptage`

- Fix issue #42 when id_quest was stored in numeric columns in Excel.

- Fix issue #44 when dms == 0

# evavelo 0.4.0

- Create geocoding functions `geocode_table_communes()`, `geocode_cities` and `geocode_cities_cp()`

- Create distance computation function `calc_distance()` and include in `process_evavelo()`

# evavelo 0.3.0

- Add button in the UI to download full file and logs

- Add update_wb() function

# evavelo 0.2.0

- #21 Add check on similarity between sheets and their _post_traitement equivalent.

- Fix issue #22 on `correct_categ.R` with identical responses in `categorie`

- Fix issues #23, #25

# evavelo 0.1.0
 
- Add "coherence" in category correction
 
# evavelo 0.0.0.9000

-   Added a `NEWS.md` file to track changes to the package.


R/read_evavelo.R
#' Read and clean eva-velo xlsx file
#'
#' Read specific sheets from an xlsx object and perform some basic cleaning.
#'
#' @param file xlsx file, Workbook object or URL to xlsx file
#'
#'
#' @return an evadata object which is in fact a list of data.frames
#' @export

read_evavelo <- function(file){
  out <- list(
    calendrier = read_calendrier(file),
    table_communes = read_table_communes(file),
    comptage_init = read_comptage(file, init = TRUE),
    comptage = read_comptage(file),
    enquete_init = read_enquete(file, init = TRUE),
    enquete = read_enquete(file),
    comptages_automatiques = read_compt_auto(file)
  )

  ## Check if any DF is empty (except for comptages_automatiques)
  empty_dfs <- out %>%
    purrr::keep(is.data.frame) %>%
    purrr::map_lgl(is_empty_df)

  if(any(empty_dfs)){
    stop(
      "Des onglets n\'ont aucune donn\u00e9es"
    )
  }

  ## Add attributes to output (S3 Object)
  class(out) <- c("evadata", class(out))
  attr(out, "geocoded") <- FALSE
  out

}

#' Read and clean "table_communes" information
#'
#' Read a specific sheet of an xlsx object and perform some basic cleaning.
#'
#' @param file xlsx file, Workbook object or URL to xlsx file
#' @param sheet Name of the worksheet containing "table_communes" information.
#'
#' @return a data.frame
#' @keywords internal
read_table_communes <- function(file, sheet = "table_communes"){

  openxlsx::read.xlsx(file,
                      sheet,
                      startRow = 2) %>%  # We skip the first row that contains global information and not data.
    janitor::clean_names() %>%
    warning_empty_sheet(sheetname = sheet)
  # geocode_table_communes()

}



#' Read and clean "comptage" information
#'
#' Read a specific sheet of an xlsx object and perform some basic cleaning.
#'
#' @param file xlsx file, Workbook object or URL to xlsx file
#' @param init boolean to import the initial sheet instead of the _post_traitement
#'
#' @importFrom rlang .data
#'
#' @return a data.frame
#' @keywords internal
read_comptage <- function(file, init = FALSE){
  sheet <- ifelse(init, "comptages_manuels", "comptages_man_post_traitements")
  comptage <- openxlsx::read.xlsx(file, sheet)

  comptage %>%
    dplyr::select(dplyr::starts_with("[")) %>%  ## Don't take in account "old" col names that could create duplicated entries
    janitor::clean_names() %>%
    dplyr::mutate(
      dplyr::across(
        dplyr::starts_with("categorie"),
        as.character
      ),
      dplyr::across(
        dplyr::starts_with("categorie_visuelle"),
        ~ stringr::str_remove(.x, "s$") ## Remove s from plurar (common user mistake)
      )
    ) %>%
    dplyr::mutate(
      date_enq = openxlsx::convertToDate(.data$date_enq),
      id_quest = as.character(.data$id_quest)
    ) %>%
    warning_empty_sheet(sheetname = sheet)
}

#' Read and clean "enquete" information
#'
#' Read a specific sheet of an xlsx object and perform some basic cleaning.
#'
#' @param file xlsx file, Workbook object or URL to xlsx file
#' @param init boolean to import the initial sheet instead of the _post_traitement
#'
#' @importFrom rlang .data
#'
#' @return a data.frame
#' @keywords internal
read_enquete <- function(file, init = FALSE) {
  sheet <- ifelse(init, "enquetes_saisies", "enquetes_post_traitement")
  enquete <- openxlsx::read.xlsx(file, sheet)
  enquete <- enquete %>%
    janitor::clean_names() %>%
    dplyr::mutate( ## Deal with empty case that are considered as logical after import
      dplyr::across(dplyr::starts_with("type_"), as.character),
      dms = as.numeric(.data$dms)
    ) %>%
    dplyr::mutate(
      type_sortie = dplyr::case_when(
        type_sortie == "Demi journ\u00e9e" ~ "Demi-journ\u00e9e",
        type_sortie == "La journ\u00e9e" ~ "Journ\u00e9e",
        TRUE ~ type_sortie)
    ) %>%
    dplyr::mutate(
      type_trajet = dplyr::case_when(
        stringr::str_detect(type_trajet, "simple") ~ "Trajet simple",
        stringr::str_detect(type_trajet, "retour") ~ "Aller-retour",
        TRUE ~ type_trajet)
    ) %>%
    dplyr::mutate(
      date_enq = openxlsx::convertToDate(.data$date_enq),
      heure_enq = as.integer(.data$heure_enq),
      revenu = as.numeric(revenu),
      id_quest = as.character(.data$id_quest),
      pays_res = as.character(.data$pays_res),
      dms = dplyr::if_else(.data$dms != 0,.data$dms, NA_real_) ## Fix issue #44 when dms = 0
    ) %>%
    dplyr::mutate( ## Avoid leading or trailing whitespaces.
      dplyr::across(
        dplyr::all_of(c("nom_site_enq", "iti_arrivee_itineraire", "iti_depart_itineraire",
                        "ville_heb", "ville_res")),
        stringr::str_trim
      )
    ) %>%
    warning_empty_sheet(sheetname = sheet)

  enquete

}

#' Read and clean "calendrier" information
#'
#' Read a specific sheet of an xlsx object and perform some basic cleaning.
#'
#' @param file xlsx file, Workbook object or URL to xlsx file
#' @param sheet Name of the worksheet caontaining "calendrier" information.
#'
#' @return a data.frame
#' @keywords internal
read_calendrier <- function(file, sheet = "calendrier_sites"){

  calendrier <- openxlsx::read.xlsx(file,
                                    sheet,
                                    startRow = 2) %>%  # We skip the first row that contains global information and not data.
    dplyr::mutate(
      date_enq = openxlsx::convertToDate(.data$date_enq)
    ) %>%
    janitor::clean_names() %>%
    warning_empty_sheet(sheetname = sheet)

}

#' Read and clean "comptages_automatiques" information
#'
#' Read a specific sheet of an xlsx object and perform some basic cleaning.
#' Compute predictors that are later user for clustering classification
#'
#' @param file xlsx file, Workbook object or URL to xlsx file
#' @param sheet Name of the worksheet containing "comptages_automatiques" information.
#'
#' @return a data.frame
#' @export
read_compt_auto <- function(file, sheet = "comptages_automatiques"){
  # Load header data ----------------------------------------------------------------------------
  header_data <- openxlsx::read.xlsx(file,
                                     sheet,
                                     sep.names = " ",
                                     rows = 1:4)

  if(is.null(header_data)){
    warning("Aucune donn\u00e9e pr\u00e9sente dans l\'onglet comptages_automatiques")
    return(NULL)
  }

  header_data <- header_data %>%
    dplyr::select(-(1:9)) %>% ## start col is 10
    t() %>%
    dplyr::as_tibble(rownames = "site_name") %>%
    dplyr::rename_with(~ c("site_name", "id_site","id_channel", "name")) %>%
    dplyr::mutate(dplyr::across(where(is.character),
                                stringr::str_squish)) ## Remove repeated white spaces

  ##TODO add a test for uniqueness of id_channel.
  if(anyDuplicated(header_data$id_channel) != 0){
    warning("id_channels dupliqu\u00e9s dans l\'onglet comptages_automatiques")
    return(NULL)
  }


  # Load data -----------------------------------------------------------------------------------
  load_data <- openxlsx::read.xlsx(file,
                                   sheet,
                                   sep.names = " ",
                                   startRow = 4)

  if(nrow(load_data) == 0){
    warning("Aucune donn\u00e9e pr\u00e9sente dans l\'onglet comptages_automatiques")
    return(NULL)
  }

  load_data <- load_data%>%
    dplyr::select(-dplyr::any_of(c("date", "annee", "mois", "jour", "type_jour"))) %>%
    dplyr::rename(date = 1) %>%
    ## Remove non-existing date (like winter> summer time) before conversion
    dplyr::filter(dplyr::if_any(c(where(is.numeric), -date),
                                ~ !is.na(.x))) %>%
    ## Transform "x" to TRUE or FALSE
    dplyr::mutate(dplyr::across(where(is.character),
                                ~ !is.na(.x))) %>%
    ## Create useful variables
    dplyr::mutate(
      date = as.POSIXct(date*3600*24, tz = "GMT", origin = "1899-12-30"), ## to avoid BUG when converting long data using openxlsx
      week_end = lubridate::wday(.data$date) %in% c(1,7),
      july_august = lubridate::month(.data$date) %in% 7:8
    ) %>%
    tidyr::pivot_longer(where(is.numeric),
                        names_to = "name",
                        values_to = "count"
    ) %>%
    dplyr::left_join(header_data,
                     by = "name")

  # Check data produced to avoid inconsistency later on
  if(all(!c("date", "jour_ferie", "pont", "vacances", "week_end", "july_august",
            "name", "count", "site_name", "id_site", "id_channel") == names(load_data))){
    warning("Probl\u00e8me de format dans l\'onglet comptages_automatiques. Impossible d\'importer")
    return(NULL)
  }

  # Compute predictors --------------------------------------------------------------------------

  pred <- load_data %>%
    dplyr::group_by(site_name, id_site, id_channel, name) %>%
    dplyr::summarize(
      ## Weekday proportion excluding holiday (working period)
      pred_wd_wp = sum_prod(!vacances, !week_end, !pont, !jour_ferie, count) /
        sum_prod(!vacances, !jour_ferie, count),
      ## Weekday proportion during holiday
      pred_wd_ho = sum_prod(vacances, !week_end, count) / sum_prod(vacances, count),
      ## July August over total
      pred_jul_aug = sum_prod(july_august, count) / sum_prod(count),
      ## Pont 14 juillet et 15 aout dans fréquentation mois aout juillet
      pred_pont_jul_aug = sum_prod(july_august, pont, count) / sum_prod(july_august, count),
      ## Proportion of count after 17:00 and before 9:00 (on weekday / working period)
      pred_wp_17_9 = sum_prod(!dplyr::between(lubridate::hour(date),9,17), !vacances, !week_end, !pont, !jour_ferie, count) /
        sum_prod(!vacances, !week_end, !pont, !jour_ferie, count),
      ## Proportion of counts between 9h and 11h in week-end
      pred_we_09_11 = sum_prod(dplyr::between(lubridate::hour(date),9,11), week_end, count) /
        sum_prod(week_end, count),
      ## Proportion of missing data
      missing_perc = sum(is.na(count)) / dplyr::n(),
      n_missing_days = dplyr::n_distinct(as.Date(date)) - dplyr::n_distinct(as.Date(date[!is.na(count)])),
      .groups = "drop"
    )

  return(pred)
}

## Utils function to check for empty sheet and create a warning
warning_empty_sheet <- function(df, sheetname){
  if(is_empty_df(df))
    warning("L\'onglet ", sheetname, " n\'a aucune donn\u00e9e.",
            immediate. = TRUE,
            call. = FALSE)
  return(df)
}

